horizons/__init__.py

```diff
@@ -0,0 +1,281 @@
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+import dateutil.parser
+from common.minimodel import Target, TargetTag, Site
+import numpy as np
+import numpy.typing as npt
+import requests
+from typing import Tuple
+
+import os
+
+@dataclass
+class EphemerisCoordinates:
+    """
+    Both ra and dec are in radians.
+
+    """
+    ra: npt.NDArray[float]
+    dec: npt.NDArray[float]
+    time: npt.NDArray[float]
+
+    @staticmethod
+    def dms2deg(dms):
+
+        if dms is None:
+            return None
+
+        d = float(dms[0])
+        m = float(dms[1])
+        s = float(dms[2])
+        sign = dms[3]
+        dd = d + m / 60. + s / 3600.
+
+        if sign == '-':
+            dd *= -1.
+
+        return dd
+
+    @staticmethod
+    def dms2rad(dms):
+        if dms is None:
+            return None
+        
+        dd = EphemerisCoordinates.dms2deg(dms)
+        rad = dd * np.pi / 180.
+        return rad
+
+    @staticmethod
+    def hms2rad(hms):
+        if hms is None:
+            return None
+
+        h = float(hms[0])
+        m = float(hms[1])
+        s = float(hms[2])
+        hours = h + m / 60. + s / 3600.
+        rad = hours * np.pi / 12.
+
+        return rad
+
+
+class HorizonsClient:
+    """
+    API to interact with the Horizons service
+    """
+
+    # A not-complete list of solar system major body Horizons IDs
+    bodies = {'mercury': '199', 'venus': '299', 'mars': '499', 'jupiter': '599', 'saturn': '699',
+               'uranus': '799', 'neptune': '899', 'pluto': '999', 'io': '501'}
+
+    def __init__(self,
+                 site: Site,
+                 path: str = 'horizons/data',
+                 start: datetime = None,
+                 end: datetime = None):
+        self.path = path
+        self.url = 'https://ssd.jpl.nasa.gov/horizons_batch.cgi'
+        self.start = start
+        self.end = end
+        self.site = site
+     
+    
+    @staticmethod
+    def generate_horizons_id(designation: str) -> str:
+        des = designation.lower()
+        return HorizonsClient.bodies[des] if des in HorizonsClient.bodies else designation
+    
+    @staticmethod
+    def _angular_distance(ra1: float, dec1: float, ra2: float, dec2: float) -> float:
+        """
+        Calculate the angular distance between two points on the sky.
+        based on https://github.com/gemini-hlsw/lucuma-core/blob/master/modules/core/shared/src/main/scala/lucuma/core/math/Coordinates.scala#L52
+        """
+        φ1 = dec1
+        φ2 = dec2
+        delta_φ = dec2 - dec1
+        delta_λ = ra2 - ra1
+        a = np.sin(delta_φ / 2)**2 + np.cos(φ1) * np.cos(φ2) * np.sin(delta_λ / 2)**2
+        return 2 * np.arctan2(np.sqrt(a), np.sqrt(1 - a))
+
+    @staticmethod
+    def interpolate(ra1: float, dec1: float, ra2: float, dec2: float, f: float) -> float:
```

Comment:
Also, this is another method we could make:
`Union[npt.NDArray[float], float]`
for all the types and `numpy` will handle both cases automatically for us.